### DISCUSSION TITLE
TVaR

### DISCUSSION THREAD

**Discussion by Quafadas** [dragonfly-ai/slash#13] [Q&A]
Here is a problem, that I want to be able to use `Vec`, `Matrix` for. 

I'm modelling a financial instrument. I have 10,000 potential outcomes for each instrument. 

I can write an extension method, to fairly easily calculate the [tVar](https://en.wikipedia.org/wiki/Tail_value_at_risk) of each instrument, and a portfolio of instruments. 

```
    def tVar(alpha: Double) : Double = {
        val numYears = thisVector.dimension
        val nte= numYears * (1.0-alpha);
        val fte= Math.floor(nte).toInt;
        val sorted = thisVector.sort
        var i = 0
        var tailSum = 0.0;
        while (i < fte) {
            tailSum += sorted(i);
            i += 1;
        }
        tailSum / fte.toDouble;
    }
```
TVaR has the advantage of being `coherent` (I think is the right word), which is a fancy way of saying we can decompose it. 

Now, I want to find the _contribution_ of each instrument to the tVar of my portfolio. 

|Inst 1|inst 2| Inst 3 | Portfolio |
|-|-|-|-|
|1|2|1|4|
|2|4|1|7|
|3|3|0|6|
|5|6|0|11|

In this simple case, the 50% (<--- normally we want a more extreme percentile, like 99%) tVar of the portoflio is 7 + 11 / 2 = 9. 

The contribution of the instruments are 
Inst 1 = (2 + 5 ) / 2 = 3.5
Inst 2 = (4 + 6) / 5 = 5
Inst 3 = 1 / 2 = 0.5

Currently, I think calculating that decomponsition isn't possible with Vec / martix?

The obvious solution would be to identify the rows of interest by returning a `NArray[Boolean] ` of the rows of interest, extracting out the sub-matrix / vectors... and from there it's easy. 

Woudl you welcome a PR which attempted this?

**Comment by dragonfly-ai** [comment]
I think you're asking about adding support for a kind of 'decomposition' the details of which resemble slicing?

**Reply by Quafadas** [reply]
Yes - I don't know how to do this reasonably, without some ability to `slice` or extract some arbitrary subset of a vector / matrix...

If we can agree a design ... I"m happy to put up a PR...

**Reply by Quafadas** [reply]
Ideas : 

`type VecIdx[N] = NArray[Boolean]`

Then an extension method which extracts that ... perhaps VecIdx could provide the same runtime constant to things it slices. That might get around the followup dimension check problem?

**Comment by Quafadas** [comment]
To support the discussion. https://github.com/dragonfly-ai/slash/pull/14

**Comment by dragonfly-ai** [comment]
I recently wrote tests for the`subMatrix` methods, which may interest you:

https://github.com/dragonfly-ai/slash/blob/main/slash/shared/src/main/scala/slash/matrix/Matrix.scala#L332

Also, the most recent release of SLASH, 0.3, added support for `MatrixSpace`:
https://github.com/dragonfly-ai/slash/blob/main/slash/shared/src/main/scala/slash/matrix/MatrixSpace.scala

The `MatrixSpace` class consists of two instances of `VectorSpace` and can be used to consolidate common runtime dimension types in one fixed path.
